matrix/common/s3_zarr_store.py

```diff
@@ -0,0 +1,188 @@
+            num_rows: (int) number of rows to write from input dataframes
+        """
+        self.exp_df = exp_df
+        self.qc_df = qc_df
+
+        # Figure out which rows of the output table this filtered chunk will be assigned.
+        output_start_row_idx, output_end_row_idx = self._get_output_row_boundaries(num_rows)
+        # Based on that, determine which zarr chunks we need to write to
+        output_start_chunk_idx, output_end_chunk_idx = self._get_output_chunk_boundaries(
+            output_start_row_idx,
+            output_end_row_idx
+        )
+
+        # Now iterate through each chunk we're supposed to write to, and write the
+        # appropriate rows to each one.
+        written_rows = 0
+        for chunk_idx in range(output_start_chunk_idx, output_end_chunk_idx):
+            output_chunk_start = chunk_idx * self._cells_per_chunk
+
+            # Get the start and end rows in the filtered matrix that correspond to
+            # this chunk as well as the start and end rows in the chunk.
+            input_row_start = int(written_rows)
+            output_row_start = int(max(0, output_start_row_idx - output_chunk_start))
+            input_row_end = int(min(output_end_row_idx - output_start_row_idx,
+                                    input_row_start + self._cells_per_chunk - output_row_start))
+            output_row_end = int(output_row_start + input_row_end - input_row_start)
+            print(f"Writing {input_row_start}:{input_row_end} --> {output_row_start}:{output_row_end}")
+
+            input_bounds = (input_row_start, input_row_end)
+            output_bounds = (output_row_start, output_row_end)
+            self._write_row_data_to_results_chunk(chunk_idx, input_bounds, output_bounds)
+            row_count = input_row_end - input_row_start
+            written_rows += row_count
+
+    def _write_row_data_to_results_chunk(self, chunk_idx: int, input_bounds: tuple, output_bounds: tuple):
+        """Write bounded row data from input to specified results chunk.
+
+        Input:
+            chunk_idx: (str) index corresponding with index of chunk in s3 zarr store
+            input_bounds: (tuple) Beginning and end of boundaries for input rows
+            output_bounds: (tuple) Beginning and end of boundaries in output rows
+        """
+        for dset in ["expression", "cell_metadata", "cell_id"]:
+            if dset == "expression":
+                values = self.exp_df.values
+            elif dset == "cell_metadata":
+                values = self.qc_df.values
+            elif dset == "cell_id":
+                values = self.exp_df.index.values
+            full_dest_key = f"s3://{self._results_bucket}/{self._request_id}.zarr/{dset}/{chunk_idx}.0"
+            print(f"Writing {dset} to {full_dest_key}")
+            if values.ndim == 2:
+                chunk_shape = (self._cells_per_chunk, values.shape[1])
+            else:
+                chunk_shape = (self._cells_per_chunk,)
```

Comment:
Is this intentional?

Reply:
Asked Marcus. This does seem weird. Missed on the first pass